File: Hephaestus/src/docker_tools.py
"""
Docker Tools - инструменты для управления Docker контейнерами.
Список, запуск, остановка, логи, exec команды.
"""
from __future__ import annotations

import json
import subprocess
from dataclasses import dataclass


@dataclass
class ToolResult:
    """Результат выполнения инструмента."""
    success: bool
    output: str
    error: str = ""


class DockerTool:
    """Управление Docker контейнерами."""

    def __init__(self, timeout: int = 30):
        self.timeout = timeout

    def list_containers(self, all_containers: bool = False) -> ToolResult:
        """Список контейнеров.

        Args:
            all_containers: Показать все контейнеры (включая остановленные)

        Returns:
            ToolResult со списком контейнеров
        """
        try:
            cmd = ["docker", "ps", "--format", "{{.ID}}\t{{.Names}}\t{{.Status}}\t{{.Image}}"]
            if all_containers:
                cmd.append("-a")

            result = subprocess.run(
                cmd,
                capture_output=True,
                text=True,
                timeout=self.timeout
            )

            if result.returncode != 0:
                return ToolResult(
                    success=False,
                    output="",
                    error=f"Docker error: {result.stderr}"
                )

            output = "ID\t\tName\t\tStatus\t\tImage\n"
            output += "=" * 80 + "\n"
            output += result.stdout

            return ToolResult(success=True, output=output)

        except FileNotFoundError:
            return ToolResult(
                success=False,
                output="",
                error="Docker not found. Install Docker first."
            )
        except subprocess.TimeoutExpired:
            return ToolResult(
                success=False,
                output="",
                error="Command timeout"
            )
        except Exception as e:
            return ToolResult(
                success=False,
                output="",
                error=f"Error: {str(e)}"
            )

    def list_images(self) -> ToolResult:
        """Список Docker образов."""
        try:
            result = subprocess.run(
                ["docker", "images", "--format", "{{.Repository}}:{{.Tag}}\t{{.ID}}\t{{.Size}}"],
                capture_output=True,
                text=True,
                timeout=self.timeout
            )

            if result.returncode != 0:
                return ToolResult(
                    success=False,
                    output="",
                    error=f"Docker error: {result.stderr}"
                )

            output = "Image\t\t\tID\t\tSize\n"
            output += "=" * 80 + "\n"
            output += result.stdout

            return ToolResult(success=True, output=output)

        except Exception as e:
            return ToolResult(
                success=False,
                output="",
                error=f"Error: {str(e)}"
            )

    def run_container(
        self,
        image: str,
        name: str | None = None,
        ports: dict[str, str] | None = None,
        env: dict[str, str] | None = None,
        volumes: dict[str, str] | None = None,
        detach: bool = True,
        command: str | None = None
    ) -> ToolResult:
        """Запустить контейнер.

        Args:
            image: Docker образ
            name: Имя контейнера
            ports: Маппинг портов {"8080": "80"}
            env: Переменные окружения {"KEY": "value"}
            volumes: Маппинг volumes {"/host/path": "/container/path"}
            detach: Запустить в фоне
            command: Команда для выполнения

        Returns:
            ToolResult с ID контейнера
        """
        try:
            cmd = ["docker", "run"]

            if detach:
                cmd.append("-d")

            if name:
                cmd.extend(["--name", name])

            if ports:
                for host_port, container_port in ports.items():
                    cmd.extend(["-p", f"{host_port}:{container_port}"])

            if env:
                for key, value in env.items():
                    cmd.extend(["-e", f"{key}={value}"])

            if volumes:
                for host_path, container_path in volumes.items():
                    cmd.extend(["-v", f"{host_path}:{container_path}"])

            cmd.append(image)

            if command:
                cmd.extend(command.split())

            result = subprocess.run(
                cmd,
                capture_output=True,
                text=True,
                timeout=self.timeout
            )

            if result.returncode != 0:
                return ToolResult(
                    success=False,
                    output="",
                    error=f"Failed to run container: {result.stderr}"
                )

            container_id = result.stdout.strip()
            output = f"Container started: {container_id}\n"
            if name:
                output += f"Name: {name}\n"
            output += f"Image: {image}\n"

            return ToolResult(success=True, output=output)

        except Exception as e:
            return ToolResult(
                success=False,
                output="",
                error=f"Error: {str(e)}"
            )

    def stop_container(self, container: str) -> ToolResult:
        """Остановить контейнер.

        Args:
            container: ID или имя контейнера

        Returns:
            ToolResult с результатом
        """
        try:
            result = subprocess.run(
                ["docker", "stop", container],
                capture_output=True,
                text=True,
                timeout=self.timeout
            )

            if result.returncode != 0:
                return ToolResult(
                    success=False,
                    output="",
                    error=f"Failed to stop container: {result.stderr}"
                )

            return ToolResult(
                success=True,
                output=f"Container stopped: {container}"
            )

        except Exception as e:
            return ToolResult(
                success=False,
                output="",
                error=f"Error: {str(e)}"
            )

    def remove_container(self, container: str, force: bool = False) -> ToolResult:
        """Удалить контейнер.

        Args:
            container: ID или имя контейнера
            force: Принудительное удаление (даже если запущен)

        Returns:
            ToolResult с результатом
        """
        try:
            cmd = ["docker", "rm"]
            if force:
                cmd.append("-f")
            cmd.append(container)

            result = subprocess.run(
                cmd,
                capture_output=True,
                text=True,
                timeout=self.timeout
            )

            if result.returncode != 0:
                return ToolResult(
                    success=False,
                    output="",
                    error=f"Failed to remove container: {result.stderr}"
                )

            return ToolResult(
                success=True,
                output=f"Container removed: {container}"
            )

        except Exception as e:
            return ToolResult(
                success=False,
                output="",
                error=f"Error: {str(e)}"
            )

    def logs(self, container: str, tail: int = 100, follow: bool = False) -> ToolResult:
        """Получить логи контейнера.

        Args:
            container: ID или имя контейнера
            tail: Количество последних строк
            follow: Следить за логами (не поддерживается)

        Returns:
            ToolResult с логами
        """
        try:
            cmd = ["docker", "logs", "--tail", str(tail), container]

            result = subprocess.run(
                cmd,
                capture_output=True,
                text=True,
                timeout=self.timeout
            )

            if result.returncode != 0:
                return ToolResult(
                    success=False,
                    output="",
                    error=f"Failed to get logs: {result.stderr}"
                )

            output = f"=== Logs for {container} (last {tail} lines) ===\n\n"
            output += result.stdout
            if result.stderr:
                output += "\n\nStderr:\n" + result.stderr

            return ToolResult(success=True, output=output)

        except Exception as e:
            return ToolResult(
                success=False,
                output="",
                error=f"Error: {str(e)}"
            )

    def exec_command(self, container: str, command: str) -> ToolResult:
        """Выполнить команду в контейнере.

        Args:
            container: ID или имя контейнера
            command: Команда для выполнения

        Returns:
            ToolResult с результатом команды
        """
        try:
            result = subprocess.run(
                ["docker", "exec", container, "sh", "-c", command],
                capture_output=True,
                text=True,
                timeout=self.timeout
            )

            output = result.stdout
            if result.stderr:
                output += "\n\nStderr:\n" + result.stderr

            success = result.returncode == 0

            return ToolResult(
                success=success,
                output=output,
                error="" if success else f"Command failed with exit code {result.returncode}"
            )

        except Exception as e:
            return ToolResult(
                success=False,
                output="",
                error=f"Error: {str(e)}"
            )

    def inspect(self, container: str) -> ToolResult:
        """Получить детальную информацию о контейнере.

        Args:
            container: ID или имя контейнера

        Returns:
            ToolResult с информацией
        """
        try:
            result = subprocess.run(
                ["docker", "inspect", container],
                capture_output=True,
                text=True,
                timeout=self.timeout
            )

            if result.returncode != 0:
                return ToolResult(
                    success=False,
                    output="",
                    error=f"Failed to inspect container: {result.stderr}"
                )

            # Парсим JSON и форматируем
            data = json.loads(result.stdout)[0]
            
            output = f"=== Container Info: {container} ===\n\n"
            output += f"ID: {data.get('Id', 'N/A')[:12]}\n"
            output += f"Name: {data.get('Name', 'N/A')}\n"
            output += f"Image: {data.get('Config', {}).get('Image', 'N/A')}\n"
            output += f"Status: {data.get('State', {}).get('Status', 'N/A')}\n"
            output += f"Running: {data.get('State', {}).get('Running', False)}\n"
            
            # Порты
            ports = data.get('NetworkSettings', {}).get('Ports', {})
            if ports:
                output += "\nPorts:\n"
                for container_port, host_bindings in ports.items():
                    if host_bindings:
                        for binding in host_bindings:
                            output += f"  {binding.get('HostPort')} -> {container_port}\n"

            # Volumes
            mounts = data.get('Mounts', [])
            if mounts:
                output += "\nVolumes:\n"
                for mount in mounts:
                    output += f"  {mount.get('Source')} -> {mount.get('Destination')}\n"

            return ToolResult(success=True, output=output)

        except Exception as e:
            return ToolResult(
                success=False,
                output="",
                error=f"Error: {str(e)}"
            )

File: src/docker_tools.rs
/// Docker Tools — управление Docker контейнерами

use std::process::Command;
use serde_json::Value;

use crate::tools::ToolResult;

const TIMEOUT_SECONDS: u64 = 30;

/// Выполнить docker команду с проверкой
fn run_docker(args: &[&str]) -> Result<String, String> {
    let output = Command::new("docker")
        .args(args)
        .output()
        .map_err(|e| format!("Docker not found: {}", e))?;

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        return Err(format!("Docker error: {}", stderr));
    }

    Ok(String::from_utf8_lossy(&output.stdout).to_string())
}

/// Список контейнеров
pub async fn docker_list(all_containers: bool) -> ToolResult {
    let mut args = vec!["ps", "--format", "{{.ID}}\t{{.Names}}\t{{.Status}}\t{{.Image}}"];
    if all_containers {
        args.push("-a");
    }

    match run_docker(&args) {
        Ok(stdout) => {
            let mut output = String::from("ID\t\tName\t\tStatus\t\tImage\n");
            output.push_str(&"=".repeat(80));
            output.push('\n');
            output.push_str(&stdout);
            ToolResult::success(output)
        }
        Err(e) => ToolResult::error(e),
    }
}

/// Список образов
pub async fn docker_images() -> ToolResult {
    match run_docker(&["images", "--format", "{{.Repository}}:{{.Tag}}\t{{.ID}}\t{{.Size}}"]) {
        Ok(stdout) => {
            let mut output = String::from("Image\t\t\tID\t\tSize\n");
            output.push_str(&"=".repeat(80));
            output.push('\n');
            output.push_str(&stdout);
            ToolResult::success(output)
        }
        Err(e) => ToolResult::error(e),
    }
}

/// Запустить контейнер
pub async fn docker_run(
    image: &str,
    name: Option<&str>,
    ports: Option<&str>,
    env: Option<&str>,
    volumes: Option<&str>,
    detach: bool,
    command: Option<&str>,
) -> ToolResult {
    let mut args = vec!["run"];
    if detach {
        args.push("-d");
    }
    if let Some(n) = name {
        args.push("--name");
        args.push(n);
    }
    if let Some(p) = ports {
        // Ожидается формат "8080:80,9090:9090"
        for pair in p.split(',').filter(|s| !s.is_empty()) {
            args.push("-p");
            args.push(pair.trim());
        }
    }
    if let Some(e) = env {
        for pair in e.split(',').filter(|s| !s.is_empty()) {
            args.push("-e");
            args.push(pair.trim());
        }
    }
    if let Some(v) = volumes {
        for pair in v.split(',').filter(|s| !s.is_empty()) {
            args.push("-v");
            args.push(pair.trim());
        }
    }
    args.push(image);
    if let Some(cmd) = command {
        args.extend(cmd.split_whitespace());
    }

    match run_docker(&args) {
        Ok(stdout) => {
            let container_id = stdout.trim();
            let mut output = format!("Container started: {}\n", container_id);
            if let Some(n) = name {
                output.push_str(&format!("Name: {}\n", n));
            }
            output.push_str(&format!("Image: {}\n", image));
            ToolResult::success(output)
        }
        Err(e) => ToolResult::error(e),
    }
}

/// Остановить контейнер
pub async fn docker_stop(container: &str) -> ToolResult {
    match run_docker(&["stop", container]) {
        Ok(_) => ToolResult::success(format!("Container stopped: {}", container)),
        Err(e) => ToolResult::error(e),
    }
}

/// Удалить контейнер
pub async fn docker_rm(container: &str, force: bool) -> ToolResult {
    let mut args = vec!["rm"];
    if force {
        args.push("-f");
    }
    args.push(container);

    match run_docker(&args) {
        Ok(_) => ToolResult::success(format!("Container removed: {}", container)),
        Err(e) => ToolResult::error(e),
    }
}

/// Логи контейнера
pub async fn docker_logs(container: &str, tail: usize) -> ToolResult {
    let tail_str = tail.to_string();
    let args = vec!["logs", "--tail", &tail_str, container];
    match run_docker(&args) {
        Ok(stdout) => {
            let output = format!(
                "=== Logs for {} (last {} lines) ===\n\n{}",
                container, tail, stdout
            );
            ToolResult::success(output)
        }
        Err(e) => ToolResult::error(e),
    }
}

/// Выполнить команду в контейнере
pub async fn docker_exec(container: &str, command: &str) -> ToolResult {
    let args = vec!["exec", container, "sh", "-c", command];
    match run_docker(&args) {
        Ok(stdout) => ToolResult::success(stdout),
        Err(e) => ToolResult::error(e),
    }
}

/// Информация о контейнере
pub async fn docker_inspect(container: &str) -> ToolResult {
    match run_docker(&["inspect", container]) {
        Ok(stdout) => {
            match serde_json::from_str::<Vec<Value>>(&stdout) {
                Ok(data) => {
                    if let Some(info) = data.first() {
                        let mut output = format!("=== Container Info: {} ===\n\n", container);
                        
                        if let Some(id) = info.get("Id").and_then(|v| v.as_str()) {
                            output.push_str(&format!("ID: {}\n", crate::truncate_chars(id, 12)));
                        }
                        if let Some(name) = info.get("Name").and_then(|v| v.as_str()) {
                            output.push_str(&format!("Name: {}\n", name));
                        }
                        if let Some(config) = info.get("Config") {
                            if let Some(image) = config.get("Image").and_then(|v| v.as_str()) {
                                output.push_str(&format!("Image: {}\n", image));
                            }
                        }
                        if let Some(state) = info.get("State") {
                            if let Some(status) = state.get("Status").and_then(|v| v.as_str()) {
                                output.push_str(&format!("Status: {}\n", status));
                            }
                            if let Some(running) = state.get("Running").and_then(|v| v.as_bool()) {
                                output.push_str(&format!("Running: {}\n", running));
                            }
                        }

                        // Ports
                        if let Some(network) = info.get("NetworkSettings") {
                            if let Some(ports) = network.get("Ports").and_then(|v| v.as_object()) {
                                if !ports.is_empty() {
                                    output.push_str("\nPorts:\n");
                                    for (container_port, bindings) in ports {
                                        if let Some(bindings) = bindings.as_array() {
                                            for binding in bindings {
                                                if let Some(host_port) = binding.get("HostPort").and_then(|v| v.as_str()) {
                                                    output.push_str(&format!("  {} -> {}\n", host_port, container_port));
                                                }
                                            }
                                        }
                                    }
                                }
                            }
                        }

                        // Volumes
                        if let Some(mounts) = info.get("Mounts").and_then(|v| v.as_array()) {
                            if !mounts.is_empty() {
                                output.push_str("\nVolumes:\n");
                                for mount in mounts {
                                    let source = mount.get("Source").and_then(|v| v.as_str()).unwrap_or("?");
                                    let dest = mount.get("Destination").and_then(|v| v.as_str()).unwrap_or("?");
                                    output.push_str(&format!("  {} -> {}\n", source, dest));
                                }
                            }
                        }

                        ToolResult::success(output)
                    } else {
                        ToolResult::error("No container data found".to_string())
                    }
                }
                Err(e) => ToolResult::error(format!("Failed to parse inspect output: {}", e)),
            }
        }
        Err(e) => ToolResult::error(e),
    }
}
